### PR TITLE
API Add extension points to MigrateFileTask

### PR DESCRIPTION
**Pull request by maxime-rainville**
This PR adds a few extension points to `MigrateFileTask` so the `solr` module can hook into there to know it needs to temporarely disable itself to avoid constant reindexing as the migration occurs.

The doc describes how to use the extension points.

I've also fix a parallel issue where the task would have quite prematurely if one of the File Helper was not present.

# Parent issue
* #8965

### CODE REVIEW

File: src/Dev/Tasks/MigrateFileTask.php [hunk 1]
```diff
@@ -46,85 +46,102 @@ public function run($request)
...
-
-            $this->logger->info('If the task fails or times out, run it again and it will start where it left off.');
-
-            $migrated = FileMigrationHelper::singleton()->run();
-            if ($migrated) {
-                $this->logger->info("{$migrated} File DataObjects upgraded");
             } else {
-                $this->logger->info("No File DataObjects need upgrading");
+                $this->extend('preFileMigrationSubtask', $subtask);
+                $this->logger->info('### Migrating filesystem and database records (move-files)');
```
Comment (ScopeyNZ) [inline src/Dev/Tasks/MigrateFileTask.php]: Another opportunity for your new `$subtask` variable.

```suggestion
                $this->logger->info("### Migrating filesystem and database records ({$subtask})");
```

File: src/Dev/Tasks/MigrateFileTask.php [hunk 2]
```diff
@@ -46,85 +46,102 @@ public function run($request)
...
-        if (in_array('move-thumbnails', $subtasks)) {
+        $subtask = 'move-thumbnails';
+        if (in_array($subtask, $subtasks)) {
             if (!class_exists(LegacyThumbnailMigrationHelper::class)) {
                 $this->logger->error("LegacyThumbnailMigrationHelper not found");
-                return;
-            }
+            } else {
+                $this->extend('preFileMigrationSubtask', $subtask);
+                $this->logger->info('### Migrating existing thumbnails (move-thumbnails)');
```
Comment (ScopeyNZ) [inline src/Dev/Tasks/MigrateFileTask.php]:
```suggestion
                $this->logger->info("### Migrating existing thumbnails ({$subtask})");
```

File: src/Dev/Tasks/MigrateFileTask.php [hunk 3]
```diff
@@ -46,85 +46,102 @@ public function run($request)
...
 
-        if (in_array('generate-cms-thumbnails', $subtasks)) {
+        $subtask = 'generate-cms-thumbnails';
+        if (in_array($subtask, $subtasks)) {
             if (!class_exists(ImageThumbnailHelper::class)) {
                 $this->logger->error("ImageThumbnailHelper not found");
-                return;
+            } else {
+                $this->extend('preFileMigrationSubtask', $subtask);
+                $this->logger->info('### Generating new CMS UI thumbnails (generate-cms-thumbnails)');
```
Comment (ScopeyNZ) [inline src/Dev/Tasks/MigrateFileTask.php]:
```suggestion
                $this->logger->info("### Generating new CMS UI thumbnails ({$subtask})");
```

File: src/Dev/Tasks/MigrateFileTask.php [hunk 4]
```diff
@@ -46,85 +46,102 @@ public function run($request)
...
+        if (in_array($subtask, $subtasks)) {
             if (!class_exists(FixFolderPermissionsHelper::class)) {
                 $this->logger->error("FixFolderPermissionsHelper not found");
-                return;
-            }
+            } else {
+                $this->extend('preFileMigrationSubtask', $subtask);
 
-            $this->logger->info('### Fixing folder permissions (fix-folder-permissions)');
+                $this->logger->info('### Fixing folder permissions (fix-folder-permissions)');
```
Comment (ScopeyNZ) [inline src/Dev/Tasks/MigrateFileTask.php]:
```suggestion
                $this->logger->info("### Fixing folder permissions ({$subtask})");
```

File: src/Dev/Tasks/MigrateFileTask.php [hunk 5]
```diff
@@ -46,85 +46,102 @@ public function run($request)
...
+        if (in_array($subtask, $subtasks)) {
             if (!class_exists(SecureAssetsMigrationHelper::class)) {
                 $this->logger->error("SecureAssetsMigrationHelper not found");
-                return;
-            }
+            } else {
+                $this->extend('preFileMigrationSubtask', $subtask);
 
-            $this->logger->info('### Fixing secure-assets (fix-secureassets)');
+                $this->logger->info('### Fixing secure-assets (fix-secureassets)');
```
Comment (ScopeyNZ) [inline src/Dev/Tasks/MigrateFileTask.php]:
```suggestion
                $this->logger->info("### Fixing secure-assets ({$subtask})");
```

**Comment by chillu** [comment]
> Does this need to be in 4.4?

Yes, otherwise everyone on CWP with a Solr index will have a bad time. It'll go to 4.4.1 rather than 4.4.0 though.

**Comment by chillu** [comment]
Wrong branch target, cherry picking back to `4.4` with 5b6d0946f. Sorry Robbie I didn't understand what you meant before ;)

**Pull request merged**